gb-apu/src/channel/sound_channel.rs

```diff
@@ -0,0 +1,255 @@
+use crate::{
+    channel::duty::Duty,
+    channel::length_counter::LengthCounter,
+    channel::sweep::Sweep,
+    channel::timer::Timer,
+    channel::volume_envelope::{Direction, VolumeEnvelope},
+    ChannelType,
+};
+use gb_bus::{Address, Error, FileOperation, IORegArea};
+
+#[derive(Debug)]
+pub struct SoundChannel {
+    pub channel_type: ChannelType,
+    pub sweep: Option<Sweep>,
+    pub duty: Option<Duty>,
+    pub length_counter: LengthCounter,
+    pub volume_envelope: Option<VolumeEnvelope>,
+    pub enabled: bool,
+    pub timer: Option<Timer>,
+}
+
+impl SoundChannel {
+    pub fn new(channel_type: ChannelType, handles_sweep: bool) -> Self {
+        SoundChannel {
+            sweep: if handles_sweep && channel_type == ChannelType::SquareWave {
+                Some(Sweep::default())
+            } else {
+                None
+            },
+            duty: if channel_type == ChannelType::SquareWave {
+                Some(Duty::default())
+            } else {
+                None
+            },
+            length_counter: LengthCounter::new(channel_type.clone()),
+            volume_envelope: if channel_type == ChannelType::SquareWave
+                || channel_type == ChannelType::Noise
+            {
+                Some(VolumeEnvelope::default())
+            } else {
+                None
+            },
+            enabled: false,
+            timer: if channel_type == ChannelType::SquareWave
+                || channel_type == ChannelType::WaveForm
+            {
+                Some(Timer::new(channel_type))
+            } else {
+                None
+            },
+            channel_type,
+        }
+    }
+
+    pub fn step(&mut self) {
+        if let Some(timer) = &mut self.timer {
+            if let Some(duty) = &mut self.duty {
+                let reached_zero = timer.step();
+                if reached_zero {
+                    duty.step();
+                }
+            }
+        }
+    }
+
+    pub fn length_counter_step(&mut self) {
+        self.enabled = self.length_counter.step();
+    }
+
+    pub fn sweep_step(&mut self) {
+        if let Some(sweep) = &mut self.sweep {
```

Comment:
```suggestion
        if let Some(ref mut sweep) = self.sweep {
```

I prefer this format